### ISSUE TITLE
.NET standard needs a monotonic clock that only increases when the debugger is paused.

### ISSUE THREAD

**Issue by adaviding** [dotnet/corefx#41713]
Consider that you need to implement a timeout for some asynchronous thread running inside your program.  You might choose to rely on [Environment.TickCount64](https://docs.microsoft.com/en-us/dotnet/api/system.environment.tickcount64?view=netcore-3.0) as a monotonic clock (to mark the passage of time).

Unfortunately, if you were to use a debugger to pause your program, you might discover that all of your timeouts had expired when the program was resumed.  Thus, you are prevented from debugging the "normal" behavior of your program.

It might be nice if there was perhaps an `Environment.DebuggerPausedTickCount64` property which increased monotonically, but only while the debugger had paused the program.  Thus a developer would be able to implement a special kind of monotonic clock using a trivial expression:

```cs
Environment.TickCount64 - Environment.DebuggerPausedTickCount64
```

The expression above would be a kind of monotonic clock that is robust against the usage of the debugger.  It would be useful for implementing timeouts for anything occurring asynchronously within the current process.  (Obviously anything running out of process is not stopped by the debugger so there are some cases where the clock would be inappropriate.)

**Comment by GrabYourPitchforks** [comment]
I don't know if there is any way to get that information in a reliable fashion. Your application's threads could appear paused for a variety of reasons. Perhaps somebody manually paused it through the threading APIs. Perhaps there are so many threads being managed by the system that your threads' quanta are further apart than you expected. The debugger is just one of many reasons your thread might appear not to be running.

One thing you could do as a mitigation is to wrap any `CancellationToken` with a helper which queries `Debugger.IsAttached`, and if that returns _true_ then delay a bit before triggering any linked token. An untested sample of how to do this follows.

```cs
public static CancellationToken WrapToken(this CancellationToken token) {
    if (!token.CanBeCanceled)
    {
        return token;
    }
    
    CancellationTokenSource source = new CancellationTokenSource();
    
    token.Register(async state => {
        while (Debugger.IsAttached)
        {
            await Task.Delay(30000);
        }
        
        ((CancellationTokenSource)state).Cancel();
    }, source, false);
    
    return source.Token;
}
```

**Comment by Clockwork-Muse** [comment]
I think the concern was specific to the debugging case.

Another alternative would be to set breakpoints before timeout checks, and manipulate the stored "start" value.

Yet another alternative would be to have your constructor take a tick `Func<long>`, which would mean you could substitute whatever kind of ticking behavior you wanted.  It would also make tests in general easier for the same reason.

**Comment by adaviding** [comment]
Full disclosure:  I don't know anything about the implementation of the .NET Runtime, or debuggers; but I feel like there would be an ABI layer (or set of entry points) that any debugger would use to pause and unpause a process (just guessing).  Or perhaps it would be an ABI in the underlying OS.  In any case, it seems possible to add something into that layer which keeps track of when the debugger had paused and resumed a process such that a value like `Environment.DebuggerPausedTickCount64` could be computed.

If you are saying that no such layer exists, then I agree that it would be too difficult to implement something like `Environment.DebuggerPausedTickCount64`.

**Comment by GrabYourPitchforks** [comment]
@jeffschwMSFT do you have suggestions for who might be more knowledgeable about this from a debugger perspective?

**Comment by jeffschwMSFT** [comment]
A challenge with such a mechanism is it would only affect some class of timeouts and not others.  What happens if the timeout were related to a physical connection to a web socket, or resource connection (eg. sql)?  There is no reliable way to fully get a debugger aware time out.

Adding @noahfalk @tommcdon

**Comment by adaviding** [comment]
@jeffschwMSFT  You are correct.  However, there are also a number of scenarios where you are using a timeout to measure the performance of something running on threads within the same process.  In such a case it would be useful, and it would always be up to a programmer's discretion as to whether they should use `Environment.DebuggerPausedTickCount64` or not.

**Comment by noahfalk** [comment]
It is messier than it may appear : ) In addition to the issue @jeffschwMSFT mentioned there are others:
- Not all debuggers work the same way. Some debuggers coordinate with the OS, some coordinate with the runtime, and some coordinate with both OS and runtime under different circumstances (interop debugging in VS). When the debugger coordinates with the OS, the runtime is just as in the dark about the pause as your managed code is. There is no callback from the kernel that warns apps when a debugger is suspending or resuming them.
- Debuggers don't have to stop processes atomically, they can start/stop individual threads. Sometimes this is used to run function evaluations while most other threads are stopped, other times it is the reverse where most threads are running but the user requested certain threads not to run.

Most apps that I have seen get around timeout issues either by avoiding them (for example the operation continues until the user clicks a cancel button in the UI to give up), or the devs add a config mechanism to disable the timeouts. The config could be very simple such as a static property in the app that can be changed using the watch window at the beginning of the debug session.

**Comment by adaviding** [comment]
Based on the info provided by @noahfalk, I should close this feature request because the implementation is infeasible at this time.
